### ISSUE TITLE
Create hooks for Automation testing

### ISSUE THREAD

**Issue by thatguynamedandy** [BrandwatchLtd/axiom#117]
We have had a few scenarios where automation tests need to hook into Axiom components. In the past the class attribute has been used, however this does make the hooks prone to change and the tests can become brittle.

One convention we have used in the past is for custom attributes e.g. `data-tid="author-card"`.  Is this something we should adopt for Axiom? a few things to consider..

- What should the attribute be?
- Does the attribute need to be unique on the page, (if so Axiom couldn't control this).
- Which elements would need attributes.
- In the future could each Component / Composite have a corresponding Automation Test class to drive it.
- Are we worried about the data attributes being output on live.

**Comment by siggysamson** [comment]
So far all of us agreed that we shouldn't use class names to target elements. We have set up all Axiom components in a way that they simply pass through all data attributes so I'd like to see a recommendation for people to simply use those.

Besides that I don't think Axiom should make any assumptions to provide anything for tests. We should keep this open to the "user". I also can't see a way how we would add a class to each component by itself without resulting in "conflicts" / duplicate class names.

I'm not worried about data attributes being in the production build. They don't reveal any insights and they don't results in a massive increase in file size (well, assuming you don't go crazy with it 😄 ). If this turns out to be an issue anyone could come up with some kind of pre/post-processing step to remove those.

**Comment by benSlaughter** [comment]
My feeling on test IDs are that they would be something I would love to see in all FE applications to  improve automation. Not only to make it easier to automate, but also allow automation to be written earlier, as the writer can have knowledge of a selector before or during development.

Data attributes to me look to be the best way to go, I think that they should not only be page unique, but if at all possible, application unique, as this means that each identifier(data attribute) can be linked to the element it identifies. The only issue with this is having multiple elements of the same type, for instance several X components. In this case I think the test will have to handle that, If i create 3 X components, I can find 3 X component selectors. 

As for which elements needs identifiers, it will have to be anything that we interact with (buttons, fields, etc) and then anything deemed important enough to require one. This can be discussed within the team, but i feel this would be similar to as mentioned above, components etc.
I think there may be some benefit in dialogs having identifiers and maybe pages. this can be used to identify when a page/dialog has been opened/closed to keep smooth tests, as this can help alleviate race conditions.

**Comment by HHogg** [comment]
Collaborating with the different engineering teams and QA engineers, we've come up with the solution merged in #278. Essentially data attributes with 12 alphanumeric ids for the ATs to target, added when and where they are needed. Axiom having it's own attributes to target elements where the components props are not spread onto (see [`<Login>`](https://github.com/BrandwatchLtd/axiom/blob/master/src/composites/login/Login.js#L91) for example).

So to answer the questions in the original question. 

> What should the attribute be?

We decided on `data-{short_project_id}-at`, for example 

```html
data-ax-at={ ID }
data-au-at={ ID }
data-viz-at={ ID }
```
> Does the attribute need to be unique on the page, (if so Axiom couldn't control this).

Unique situations can be handled by combining project IDs. for example 

```jsx
<Page data-au-at={ ID }>
  <Login .... />
</Page> 

// Username input can be targeted by "[data-au-at={ ID }] [data-ax-at={ axiomAtIds.Login.username }]"
```
> Which elements would need attributes.

Add them where they are required. 

> In the future could each Component / Composite have a corresponding Automation Test class to drive it.

This related more to #218, but this would be awesome 😸 

> Are we worried about the data attributes being output on live.

Because of the need to run ATs on Live, this would be unavoidable. But there are options that these could be stripped for certain images, for example using a babel task.
